src/click/parser.py

```diff
@@ -337,11 +337,27 @@ def _match_long_opt(self, opt, explicit_value, state):
             if explicit_value is not None:
                 state.rargs.insert(0, explicit_value)
 
+            # if the option is a prompt and `prompt_required` is false,
+            # then the current rarg will serve as a flag, meaning the
+            # user will be prompted for an input if no argument is passed.
             nargs = option.nargs
             if len(state.rargs) < nargs:
-                _error_opt_args(nargs, opt)
```

Comment:
I've noticed this block of code in `_match_long_opt` is identical to the one in `_match_short_opt`. So when making changes, I kept it this way and added the changes twice. Should this be factored out or is it not too big of a deal?